Add OS detection for Windows 11 and Server 2022

mikezhang1234567890 opened the pull request:
The update to Windows 11 does not increment the major version,
so we detect using other info such as CurrentBuild number.

Signed-off-by: Mike Zhang <mike.h.zhang@ibm.com>

Fixes https://github.com/eclipse-openj9/openj9/issues/13755#issuecomment-951230263

Can you take a look please @pshipton

mikezhang1234567890 commented:
One thing is that the end result string can be "Windows Server 2022 Business" or "Windows 10 Enterprise" since some of the cases go to the branch that uses the `ProductName` registry key. Is the extra word a problem for the result?

pshipton commented on port/win32/omrsysinfo.c:
```diff
@@ -520,7 +530,15 @@ WIN32_WINNT version constants :
 							} else if (0 == strcmp(ridBuffer, "1809")) {
 								PPG_si_osType = "Windows Server 2019";
-							} else {
+							} else if (0 == strcmp(ridBuffer, "2009")) {
```
Let's do the same as OpenJDK and check for build 20347 or later.

pshipton commented on port/win32/omrsysinfo.c:
```diff
@@ -535,6 +553,23 @@ WIN32_WINNT version constants :
+				/* Windows 11 has same major and minor version as 10, but starts at current build number 22000 */
+				if ((defaultTypeName == PPG_si_osType) && !isServerMajorVersion10 && IsWindows10OrGreater()
+					&& (ERROR_SUCCESS == RegQueryValueExA(hKey, CURRENT_BUILD_KEY, NULL, NULL, NULL, &valueSize))) {
+					char currentbuildBuffer[6]; /* CurrentBuild values have been 5 digits so far */
```
There should be some extra space so this doesn't overflow and skip the following code if a larger build number occurs in the future.

pshipton commented on port/win32/omrsysinfo.c:
```diff
@@ -535,6 +553,23 @@ WIN32_WINNT version constants :
+					char currentbuildBuffer[6]; /* CurrentBuild values have been 5 digits so far */
+					if (sizeof(currentbuildBuffer) >= valueSize) {
+						if (ERROR_SUCCESS == RegQueryValueExA(hKey, CURRENT_BUILD_KEY, NULL, NULL, (LPBYTE)currentbuildBuffer, &valueSize)) {
+							if (0 == strcmp(currentbuildBuffer, "22000")) {
```
This should convert the string to an integer so it can check >= 22000. We want it to work for future updates of Windows 11.

pshipton commented:
> One thing is that the end result string can be "Windows Server 2022 Business" or "Windows 10 Enterprise" since some of the cases go to the branch that uses the ProductName registry key. Is the extra word a problem for the result?

Better to match the OpenJDK behavior and stick with the basic strings. We only want to go to the registry if nothing else works.

mikezhang1234567890 commented:
Changed it so it always uses CurrentBuild number to different Windows 10+ versions, using the value from `GetVersionEx` if we call that or from the registry if we're on the other path.

Also removed code looking at the registry key ReleaseId in favour of using CurrentBuild for detecting Server 2016 and 2019 in addition to the Windows 11 based 2022.

Tested against Windows 11 and 10 and the server equivalents, but only with a JDK compiled on a Server 2016 machine.

Can you take another look @pshipton ?

pshipton commented:
> but only with a JDK compiled on a Server 2016 machine

I don't think it matters where you compile, but what tool chain was used to compile. OpenJ9 jdk8 uses VS2013 which uses one code path, and other versions (including IBM Java 8) use VS2017 or later which uses the other code path.

pshipton commented:
jenkins build win

pshipton commented:
I did some testing myself, with OpenJ9 JVMs for 8, 8 32-bit, 11, 17, and IBM Java 8 32/64. On Windows 10, 11, Server 2016, 2019, 2022. It all looked good.

pshipton commented:
@0xdaryl
